examples/pytorch/distributed_stl10/stl10_example.py

```diff
@@ -0,0 +1,360 @@
+    top5 = AverageMeter()
+
+    # switch to evaluate mode
+    model.eval()
+
+    with torch.no_grad():
+        end = time.time()
+        for i, (input, target) in enumerate(val_loader):
+
+            input = input.cuda(non_blocking=True)
+            target = target.cuda(non_blocking=True)
+
+            with torch.no_grad():
+                # compute output
+                logits = model(input)
+                loss = criterion(logits, target)
+
+            # Measure accuracy
+            prec1, prec5 = accuracy(logits.data, target.data, topk=(1, 5))
+
+            # Average loss and accuracy across processes for logging
+            reduced_loss = reduce_tensor(loss.data, world_size)
+            prec1 = reduce_tensor(prec1, world_size)
+            prec5 = reduce_tensor(prec5, world_size)
+
+            # to_python_float incurs a host<->device sync
+            batch_size = input[0].size(0)
+            losses.update(to_python_float(reduced_loss), batch_size)
+            top1.update(to_python_float(prec1), batch_size)
+            top5.update(to_python_float(prec5), batch_size)
+
+            torch.cuda.synchronize()
+            batch_time.update((time.time() - end) / log_freq)
+            end = time.time()
+
+            if i % log_freq == 0 and is_master:
+                # Only the first node should log infos.
+                print(
+                    f"Test: [{i}/{len(val_loader)}]\t"
+                    f"Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t"
+                    f"Speed {world_size * batch_size / batch_time.val:.3f} ({world_size * batch_size / batch_time.avg:.3f})\t"
+                    f"Loss {losses.val:.10f} ({losses.avg:.4f})\t"
+                    f"Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t"
+                    f"Prec@5 {top5.val:.3f} ({top5.avg:.3f})"
+                )
+
+        if is_master:
+            print(f' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}')
+
+    return top1.avg
+
+
+def run(batch_size: int,
+        epochs: int,
+        learning_rate: float,
+        log_interval: int,
+        save_model: bool):
+    # number of nodes / node ID
+    n_nodes = int(os.environ['SLURM_JOB_NUM_NODES'])
+    node_id = int(os.environ['SLURM_NODEID'])
+
+    # local rank on the current node / global rank
+    local_rank = int(os.environ['SLURM_LOCALID'])
+    global_rank = int(os.environ['SLURM_PROCID'])
+
+    # number of processes / GPUs per node
+    world_size = int(os.environ['SLURM_NTASKS'])
+    n_gpu_per_node = world_size // n_nodes
+
+    # define master address and master port
+    hostnames = subprocess.check_output(['scontrol', 'show', 'hostnames', os.environ['SLURM_JOB_NODELIST']])
+    master_addr = hostnames.split()[0].decode('utf-8')
+
+    # set environment variables for 'env://'
+    os.environ['MASTER_ADDR'] = master_addr
+    os.environ['MASTER_PORT'] = str(29500)
+    os.environ['WORLD_SIZE'] = str(world_size)
+    os.environ['RANK'] = str(global_rank)
```

Comment:
This can be simplified a little using `initfile` initialization method, but I am not sure what is the best practice.

Reply:
Environment variable initialization is the default method in torch. The initfile method can lead to unexpected behaviour if the previous run was unable to correctly clean the file and I find it a bit more complicated. If you want I can implement it.

Reply:
Also, I'm not not sure how this would work as if we use the Initfile way, we need to know the master address ahead of time.